docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,239 @@
+
+Namely, let's assume that transaction `tx_1` was sent by node `B` to node `C` which forwarded it to `A`. `A` already has the transaction and notifies node `C` about this. Node `C` will look up the senders of transaction `tx_1` and in the future disable transaction forwarding from node `B` to node `A`.
+
+In reality, if node `C` ha received `tx_1` from multiple senders (`X`,`Y`, `B`), it will pick the first one that has sent it and disable that route. 
+
+When a connection to a peer is lost a node sends a `Reset` message to its peers. It
+signals to the peers that they should delete all routing information related to this node. This assures that, in case the network topology changes, or a peer becomes unavailable, the node is able to discover transactions via a previously disabled route. 
+
+### Redundancy control
+
+In an ideal setting, receiving a transaction more than once is not needed. But in a Byzantine setting, allowing for only one route for a transaction can introduce an attack vector. 
+
+Operators can thus configure a desired redundancy and the gossip protocol will adjust route disabling based on this setting. Redundancy is defined as the ratio between duplicate and unique transactions. Operators can also define how much they tolerate the redundancy to deviate from the desired level.
+
+By default, the redundancy is set to 1 and impacts the gossiping of transactions as follows:
+
+
+```
+  redundancy := duplicateTxs / firstTimeTxs
+  if redundancy < r.redundancyLowerBound:
+    peer.send(Reset)
+  if redundancy > redundancyUpperBound
+    enableRouteDisabling() 
+```
+
+Redundancy is adjusted whenever a new transaction is added into the mempool.
+(TODO - revisit whether this should also be done on duplicates.)
+
+
+### Impacted areas of code
+
+#### Mempool reactor
+
+
+The bulk of the changes is constrained to the mempool reactor. 
+We introduce a new communication channel, called `Mempool Control Channel`. The channel
+is used to transmit the messages needed for the implementation of the protocol. 
+
+Additionally, the mempool reactor is extended with a 
+`gossipRouter` and a `redundancyControl` struct to keep track of the redundancy
+of transactions in the system. 
+Each node is keeps track of the routes it disabled between its peers. 
+
+```
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating transactions. Source
+	// and target are node IDs.
+	disabledRoutes map[nodekey.ID]p2pIDSet
+}
+``` 
+
+
+```
+type redundancyControl struct {
+	txsPerAdjustment int64
+
+	// Pre-computed upper and lower bounds of accepted redundancy.
+	lowerBound float64
+	upperBound float64
+
+	mtx          cmtsync.RWMutex
+	firstTimeTxs int64 // number of transactions received for the first time
+	duplicates   int64 // number of duplicate transactions
+
+	// If true, do not send HaveTx messages.
+	blockHaveTx atomic.Bool
+}
+
+```
+
```

Comment:
In addition to the new data structures, I would include the changes to the current code. 
- Handling of new messages in `Receive` function, in the new channel.
- Extend `TryAddTx` to count first-time and duplicate txs; reply with a `HaveTx` message if tx is in cache. 
- Filter disabled routes in `broadcastTxRoutine` (only if DOG is enabled, otherwise keep filtering senders as in the current code).
- New logic on `RemovePeer`.

Reply:
It's important to emphasize that the new code does not change the current behaviour.